### ISSUE TITLE
Policy Public To Public Mappings / Preference of headers

### ISSUE THREAD

**Issue by Leon0402** [include-what-you-use/include-what-you-use#1221]
IWYU has some interesting handling of the visibility specifier, which I don't quite understand. Based on the boost script posted in an (unfortunatly) closed Pull Request, I have the following generated rules: 

```
{ include: ["<boost/geometry/algorithms/append.hpp>", public, "<boost/geometry/geometry.hpp>", public ] },
    { include: ["<boost/geometry/algorithms/area.hpp>", public, "<boost/geometry/geometry.hpp>", public ] },
    { include: ["<boost/geometry/algorithms/assign.hpp>", public, "<boost/geometry/geometry.hpp>", public ] },
    { include: ["<boost/geometry/algorithms/buffer.hpp>", public, "<boost/geometry/geometry.hpp>", public ] },
    { include: ["<boost/geometry/algorithms/centroid.hpp>", public, "<boost/geometry/geometry.hpp>", public ] },
    { include: ["<boost/geometry/algorithms/clear.hpp>", public, "<boost/geometry/geometry.hpp>", public ] },
...
 { include: ["<boost/geometry/geometry.hpp>", public, "<boost/geometry.hpp>", public ] },
....
```

Note: I modified the script a little bit. Because `<boost/geometry/geometry.hpp>` is a facade header for all these. So I'm not entirely sure if the rules are correct here. Originally just the last line was included. 

Example 1 (assume symbols from these headers are needed):
```
#include <boost/geometry/algorithms/append.hpp>
#include <boost/geometry/algorithms/envelope.hpp>
#include <boost/geometry/geometries/linestring.hpp>
#include <boost/geometry/geometries/point.hpp>
#include <boost/geometry/geometries/polygon.hpp>
#include <boost/geometry/geometries/ring.hpp>
```

Nothing changes on iwyu run. If I add `#include <boost/geometry/geometry.hpp>` to this:

Example 2
```
#include <boost/geometry/algorithms/append.hpp>
#include <boost/geometry/algorithms/envelope.hpp>
#include <boost/geometry/geometries/linestring.hpp>
#include <boost/geometry/geometries/point.hpp>
#include <boost/geometry/geometries/polygon.hpp>
#include <boost/geometry/geometries/ring.hpp>

#include <boost/geometry/geometry.hpp>
```

Then still nothing changes. But I can now remove all headers like this manually: 

Example 3
```
#include <boost/geometry/geometry.hpp>
```

Fine for iwyu as well, it won't add any of the removed headers. The same if I use: 

Example 4
```
#include <boost/geometry.hpp>
```

instead. But what if I just remove all headers: 

Example 5
```
```

Now iwyu will add `#include <boost/geometry.hpp>`  (Note not #include <boost/geometry/geometry.hpp>, so really the most general facade header)

So my two oberservations are with public to public mapping (chains): If nothing is given it will prefer the left public header after following the complete include chain. If there are files given, they will never be replaced by another public dependency. This somehow seems consistent with some statement I read in the issues that two public headers are seen as interchangable (although it's not officially documented I think).

Some thoughts about this:
- If there is for instance already  <boost/geometry.hpp> in the include list, it should at least remove the other includes. Makes no sense to keep multiple interchangable headers (like in Example 2). Just use the most generic one.
- I wonder if this policy really is good after all. Basically this means the output of iwyu for the same files (minus differences in the includes) is not deterministic. In my opinion there are only two best outputs depending on preference perhaps:
  - Only have the most general facade header at the end. For people who prefer less includes.
  - Only have the most specific public headers and use facade header only for private headers. For people who generally dislike facade headers because of performance impacts. (Note that in the case there is one public and one private header, the ideal solution would be to just use the facade headers, because it has to be used anyways for the private header)

In my opinion it would be good if the tool always finds the best solution (taking into accounts preferences as well) and be thus deterministic. 

Ideally there would be an option, which will let iwyuo use either generate the soltuion from Example 1 or Example 5, but none of the others. Would it be a solution for iwyu to basically just remove all includes everywhere and then figures out the includes from scratch? This would lead to solution 5.

What are your thoughts about this?

**Comment by kimgr** [comment]
I would prefer for IWYU to have a canonical policy, "include what you use", but have enough flexibility in the mapping language that you can pick and choose a bit how the best header is selected.

And to some extent, I think that's already true -- if you mark the mappings for the non-facade headers as private, IWYU should naturally move toward the public header. If you remove the mappings from non-facade to facade, IWYU should move toward more granular headers.

Try to express general policies _on top of_ mappings seems hard to me. I think it's easier to just change the mappings :-)

**Comment by Leon0402** [comment]
Yeah I think you got a point that you have somewhat a choice currently with the mappings files. This is basically what I ended up with right now when specifying my own mappings. The main problem though is designing the mappings in a way that every header is either public or private. This boils down to two main problems I have at the moment: 

a) Built in mapping files and existing complex mappings files / scripts (like the one for boost). These already define some headers as public and trying to fix that is kind of hard.
b) Writing automated script that get this right is hard as well. Say I have a well known facade header and write script collecting the mappings for this automatically. It seems natural to do a private - public mapping. But what if there is another facade header for this facade header. Then I should have chosen a private - private mapping. I ran into such issues quite a lot where I got these failure assertions: Header has different visibilities. 

I was wondering if this requirement is strictly necessary. Wouldn't it be an option to say: If a header is marked as private, this takes preference? For instance say there are these two rules: 
```
 { include: ["<boost/geometry/geometry.hpp>", private, "<boost/geometry.hpp>", public ] },
 { include: ["<boost/geometry/algorithms/append.hpp>", private, "<boost/geometry/geometry.hpp>", public ] },
```
Currently the will lead to an error, because the header `<boost/geometry/geometry.hpp>` is marked private and public as well. But I wonder if there would be any problems if iwyu would treat the header in such cases as private?

**Comment by kimgr** [comment]
Yeah, or get rid of that assertion entirely, and just overwrite with a warning. See also: https://github.com/include-what-you-use/include-what-you-use/issues/414

**Comment by Leon0402** [comment]
That seems reasonable from my perspective. Although I wouldn't make it a warning, but rather an "info" or something like that. Warning still suggest that doing it like this is wrong. 

Would the assertion should need to be removed and then everything works perfectly or are further adjustments needed to actually implement the override strategy here?
Are you still waiting for input from other contributors to decide if this is the direction to go?

**Comment by bolshakov-a** [comment]
Some tweaking and discussion is here: #1990.
